### DISCUSSION TITLE
`IntoPy` not implemented for `Bound`

### DISCUSSION THREAD

**Discussion by wyfo** [PyO3/pyo3#4049] [General]
I've just updated some old code to 0.21, and I had the following code:
```rust
let list: Py<PyList> = Python::with_gil(|py| PyList::new(py, vec).into_py(py));
```
`PyList::new` is deprecated, so I've changed to
```rust
let list: Py<PyList> = Python::with_gil(|py| PyList::new_bound(py, vec).into_py(py));
```
and I got the following error `expected `Py<PyList>`, but found `PyObject`.

Why is there no `impl IntoPy<Py<PyList>> for Bound<'_, PyList>`, while an `impl IntoPy<Py<PyList>> for &PyList` exists?

**Comment by adamreichold** [comment]
Not really an answer as to why the trait impl does not exist, but I think you want [`Bound::unbind`](https://docs.rs/pyo3/latest/pyo3/prelude/struct.Bound.html#method.unbind). We also recently extended the guide on this topic, c.f. <https://github.com/PyO3/pyo3/pull/4031>.

(I think in this particular case, `unbind` is generally preferable because it is the more narrow operation and hence better self-documenting. Still for generic context, a blanket impl like `impl<T> IntoPy<Py<T>> for Bound<'_, T>` might be nice.)

**Reply by wyfo** [reply]
Indeed, I can use `unbind`, and even `into`. My point is I don't know why this impl has not been added, especially thinking about generic context. Do you think I should open an issue/PR to add this impl?

**Reply by adamreichold** [reply]
I think it would be preferable to engage in #4041 hashing out how to simplify the overall conversion trait approach.

**Reply by davidhewitt** [reply]
I think it's the second optional item in #3960 

If I recall, the problem is we already have `IntoPy<Py<PyAny>> for Bound<T>`, and this conflicts with a generic `IntoPy<Py<T>> for Bound<T>`.

Probably there's a solution like what we did for the `Deref` implementations. It would probably break type inference to ship a solution in a patch release but we could include it in 0.22

This feels also very related to what we're discussing in #4041 and generic traits.

**Reply by davidhewitt** [reply]
(oops, @adamreichold said the same thing at exactly the same time 😂)

**Reply by wyfo** [reply]
> If I recall, the problem is we already have IntoPy<Py<PyAny>> for Bound<T>, and this conflicts with a generic IntoPy<Py<T>> for Bound<T>.

I think that's why `impl IntoPy<Py<$name>> for &$name` is in `pyobject_native_type_named`. Maybe we could add `impl IntoPy<Py<$name>> for Bound<'_, $name>` and `impl IntoPy<Py<$name>> for &Bound<'_, $name>` in `pyobject_native_type_named` as well.

**Reply by davidhewitt** [reply]
Ah yes, that would probably work. I also need to have a good read of #4060 which hopefully gives us a way out of the current bundle of traits!
